### PR TITLE
Improve initialization type frames

### LINKED ISSUE #610: `@MonotonicNonNull` fields now prevent `final`-class constructors from calling a method at the end
New in 3.39.0-eisop1:

```
$ cat I.java
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;

final class I {
  @MonotonicNonNull String s;

  I() {
    init();
  }

  void init() {}
}

$ checker/bin/javac -processor nullness I.java
I.java:7: error: [method.invocation.invalid] call to init() not allowed on the given receiver.
    init();
        ^
  found   : @UnderInitialization I
  required: @Initialized I
1 error
```

I could believe that this has actually been unsound all along for reasons that I hadn't noticed, but it looks to me like it ought to continue to be permitted.

### PR DESCRIPTION
**Pull request by flo2702**
Fixes #610

An object of final class type `A` is now correctly considered `@Initialized` when all non-monotonic, non-null fields have been assigned. Previously, it was only considered `@UnderInitialization(A.class)`.

The problem only occurred if some fields of the class were either monotonic or nullable, which was not previously covered by test cases.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java [hunk 1]
```diff
@@ -381,16 +396,18 @@ protected void reportMethodInvocabilityError(
             return;
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java:396]: Can you add a comment, before the `receiverType` local variable, what this check is for?

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -101,6 +113,75 @@ public InitializationStore getExceptionalExitStore(Tree tree) {
...
 
+    /**
+     * Returns the fields that are not yet initialized in a given store, taking into account the
+     * target checker.
+     *
+     * <p>A field is initialized if
+     *
+     * <ul>
+     *   <li>it is initialized in the initialization store, i.e., has been assigned.
+     *   <li>it has a non-top qualifier that does not have the meta-annotation {@link
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/initialization/InitializationAnnotatedTypeFactory.java:124]: What does the "it has" mean here? Also the "its declared qualifier" below confused me.
Is one about the value in the store the other the declared field type?

File: checker/tests/initialization/TypeFrames5.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.initialization.qual.UnderInitialization;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class TypeFrames5 {
+    public @Nullable Object f;
+
+    public TypeFrames5(boolean dummy) {
+        @UnderInitialization(TypeFrames5.class) TypeFrames5 a = this;
```
Comment (wmdietl) [inline checker/tests/initialization/TypeFrames5.java:9]: I added this test to show that the type of `this` is correct even for non-final classes, which was previously also broken.
Can you think of another case to test?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for this fix, in particular all the documentation improvements!
I have a few small questions.
I've added a further test and renamed the PR, see whether you like these changes.

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java (hunk 1 above)
Comment (flo2702) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java:402]: Done in [aed8e24](https://github.com/eisop/checker-framework/pull/621/commits/aed8e241bca2625501d1315ebbbb82b32c816dd4)

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationAnnotatedTypeFactory.java (hunk 2 above)
Comment (flo2702) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationAnnotatedTypeFactory.java]: Clarified in [aed8e24](https://github.com/eisop/checker-framework/pull/621/commits/aed8e241bca2625501d1315ebbbb82b32c816dd4)

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java [hunk 4]
```diff
@@ -363,13 +364,27 @@ protected void reportMethodInvocabilityError(
...
         }
 
+        AnnotationMirror init = expected.getAnnotation(Initialized.class);
         AnnotationMirror unknownInit = expected.getAnnotation(UnknownInitialization.class);
         AnnotationMirror underInit = expected.getAnnotation(UnderInitialization.class);
+
+        // If the found receiver type is wrong, we still do not report an error
```
[the remark this answers is not in the document]
Comment (flo2702) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java]: Clarified in [aed8e24](https://github.com/eisop/checker-framework/pull/621/commits/aed8e241bca2625501d1315ebbbb82b32c816dd4)

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java [hunk 5]
```diff
@@ -358,48 +360,72 @@ protected void reportCommonAssignmentError(
...
         uninitializedFields.removeAll(initializedFields);
 
-        filterInitializedFields(node, uninitializedFields, true);
-
         if (!uninitializedFields.isEmpty()) {
-            StringJoiner fieldsString = new StringJoiner(", ");
-            for (VariableTree f : uninitializedFields) {
-                fieldsString.add(f.getName());
-            }
+            // TODO: improve the error message by showing the uninitialized fields
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java:418]: @Ao-senXiong Can you please open an issue for this TODO and see how to improve this error message?

**Review by wmdietl — APPROVED** [review]
Thanks for the further improvements!

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java (hunk 5 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java:418]: > @Ao-senXiong Can you please open an issue for this TODO and see how to improve this error message?

Ok, I will open up an issue

**Pull request merged**